### PR TITLE
changes to episode 3, tasks 9-23

### PR DESCRIPTION
**Pull request by mallewellyn**
In response to tasks 9-23 of #115 

In summary:

- clearly framed the linear regression section as a recap (feels a bit random where it is)
- all the talk of generalisability in the regularisation section seems unconnected to the introductory lesson motivation. Have tried to connect the two
- ridge regression section title added for consistency with lasso (maybe helps in connecting the two) and example combined into this section as reasons for restricting the models should be clear/well-motivated already in my opinion
- several changes to cross validation

### CODE REVIEW

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 1]
```diff
@@ -353,75 +364,68 @@ horvath_mat <- methyl_mat[, features]
...
+Now let's fit a linear model to our training data and calculate the training error. Here we use the mean of the squared difference between our predictions and the observed data, or "mean squared error" (MSE).
+
+```{r trainlm}
+## Fit a linear model
+# as.data.frame() converts train_mat into a data.frame
+#Using the `.` syntax above together with a `data` argument will lead to
+#the same result as using `train_age ~ train_mat`: R will fit a multivariate 
+#regression model in which each of the columns in `train_mat` is used as 
+#a predictor. We opted to use the `.` syntax because it will help us to 
+#obtain model predictions using the `predict()` function.
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:385]:
```suggestion
# Using the `.` syntax above together with a `data` argument will lead to
# the same result as using `train_age ~ train_mat`: R will fit a multivariate 
# regression model in which each of the columns in `train_mat` is used as 
# a predictor. We opted to use the `.` syntax because it will help us to 
# obtain model predictions using the `predict()` function. 
```

This is just a style thing for me, probably doesn't bother anybody else!

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 2]
```diff
@@ -552,19 +555,33 @@ or correlated predictors. As we reduce the importance of one feature, we can
 weight to another feature that represents similar information.
 
 Since we split the data into test and training data, we can prove that ridge
-regression gives us a better prediction in this case:
+regression gives us a better prediction compared to the model with no regularisation. 
+Let's generate our predictions under the ridge regression model and calculate
+the mean squared error in the test set:
 
 ```{r pred-ridge-lm, fig.width = 10, fig.cap="Cap", fig.alt="Alt"}
+##Predict using the ridge model
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:563]:
```suggestion
# Predict using the ridge model
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 3]
```diff
@@ -552,19 +555,33 @@ or correlated predictors. As we reduce the importance of one feature, we can
...
 pred_ridge <- predict(ridge_fit, newx = test_mat)
-err_ridge <- apply(pred_ridge, 2, function(col) mse(test_age, col))
-min(err_ridge)
-err_lm
 
-which_min_err <- which.min(err_ridge)
+##Calculate MSE
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:566]:
```suggestion
# Calculate MSE for every column of the prediction matrix against the vector of true ages
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 4]
```diff
@@ -552,19 +555,33 @@ or correlated predictors. As we reduce the importance of one feature, we can
...
+err_ridge <- apply(pred_ridge, 2, function(col) mse(test_age, col)) 
 min_err_ridge <- min(err_ridge)
+
+##Save some results for later analysis
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:570]: # identify the lowest MSE value out of all of the columns of the prediction matrix

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 5]
```diff
@@ -552,19 +555,33 @@ or correlated predictors. As we reduce the importance of one feature, we can
...
 
 Since we split the data into test and training data, we can prove that ridge
-regression gives us a better prediction in this case:
+regression gives us a better prediction compared to the model with no regularisation. 
+Let's generate our predictions under the ridge regression model and calculate
+the mean squared error in the test set:
 
 ```{r pred-ridge-lm, fig.width = 10, fig.cap="Cap", fig.alt="Alt"}
+##Predict using the ridge model
 pred_ridge <- predict(ridge_fit, newx = test_mat)
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:564]:
```suggestion
# obtain a prediction matrix, where every column corresponds to a different lambda value
pred_ridge <- predict(ridge_fit, newx = test_mat)
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 6]
```diff
@@ -736,35 +755,45 @@ drawplot()
...
 ```{r cvfig, echo = FALSE, fig.cap = "Schematic representiation of a $K$-fold cross-validation procedure.", fig.alt = "The data is divided into $K$ chunks. For each cross-validation iteration, one data chunk is used as the test set. The remaining $K-1$ chunks are combined into a training set."}
 knitr::include_graphics("../fig/cross_validation.png")
 ```
 
-We can use this new idea to choose a lambda value, by finding the lambda
-that minimises the error across each of the test and training splits.
+We can use this new idea to choose a lambda value by finding the lambda
+that minimises the error across each of the test and training splits. In R:
 
 ```{r lasso-cv, fig.cap="Cross-validated mean squared error for different values of lambda under a LASSO penalty.", fig.alt="Alt"}
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:778]: ``{r lasso-cv, fig.cap="Cross-validated mean squared error for different values of lambda under a LASSO penalty.", fig.alt="Mean squared error (with confidence interval) for a range of different lambda values. Mean squared error values are shown as red dots, with black error bars on the y-axis corresponding to the corresponding confidence intervals."}

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 7]
```diff
@@ -736,35 +755,45 @@ drawplot()
...
+##Fit lasso
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:779]:
```suggestion
# fit lasso model with cross-validation across a range of lambda values
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 8]
```diff
@@ -736,35 +755,45 @@ drawplot()
...
 lasso <- cv.glmnet(methyl_mat[, -1], age, alpha = 1)
 plot(lasso)
+
+##Extract the coefficients and return those that are non-zero
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:783]:
```suggestion
# Extract the coefficients corresponding to the lowest mean squared error from cross-validation
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 9]
```diff
@@ -736,35 +755,45 @@ drawplot()
...
 coefl <- coef(lasso, lasso$lambda.min)
-selected_coefs <- as.matrix(coefl)[which(coefl != 0), 1]
+selected_coefs <- as.matrix(coefl)[which(coefl != 0), 1] ; selected_coefs
```
Comment (alanocallaghan) [inline _episodes_rmd/03-regression-regularisation.Rmd:785]:
```suggestion
# select only non-zero coefficients
selection <- which(coefl != 0)
# and convert to a normal matrix
selected_coefs <- as.matrix(coefl)[selection, 1]
selected_coefs
```

**Comment by alanocallaghan** [comment]
Added some minor comments, but this deserves a proper review so will return to it later or if Ailith/Cata beat me to it, obviously happy for them to review instead

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 10]
```diff
@@ -353,73 +364,68 @@ horvath_mat <- methyl_mat[, features]
...
+> > ```
 > > 
-> > 3. The mean squared error of the model is the mean of the square of the
-> >    residuals. This seems very low here -- on average we're only off by 
-> >    about a year!
-> >    ```{r testerror}
-> >    mean(residuals(fit_horvath)^2)
-> >    ```
+> > The mean squared error for the test set is the mean of the squared 
+> > error of the predicted values and true test data.
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:415]:
```suggestion
> > error between the predicted values and true test data.
```
?

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 11]
```diff
@@ -552,19 +555,34 @@ or correlated predictors. As we reduce the importance of one feature, we can
 weight to another feature that represents similar information.
 
 Since we split the data into test and training data, we can prove that ridge
-regression gives us a better prediction in this case:
+regression gives us a better prediction compared to the model with no regularisation.
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:558]:
```suggestion
regression predicts the test data better than the model with no regularisation. 
```

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 4 above)
Comment (ailithewing) [reply _episodes_rmd/03-regression-regularisation.Rmd:571]: Made a minor edit to Alan's... same principle.

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 12]
```diff
@@ -552,19 +555,34 @@ or correlated predictors. As we reduce the importance of one feature, we can
 weight to another feature that represents similar information.
 
 Since we split the data into test and training data, we can prove that ridge
-regression gives us a better prediction in this case:
+regression gives us a better prediction compared to the model with no regularisation. 
+Let's generate our predictions under the ridge regression model and calculate
+the mean squared error in the test set:
 
 ```{r pred-ridge-lm, fig.width = 10, fig.cap="Cap", fig.alt="Alt"}
+# obtain a matrix of predictions from the ridge model,
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:563]:
```suggestion
# Obtain a matrix of predictions from the ridge model,
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 13]
```diff
@@ -552,19 +555,34 @@ or correlated predictors. As we reduce the importance of one feature, we can
...
-which_min_err <- which.min(err_ridge)
+# Calculate MSE for every column of the prediction matrix against the vector of true ages
+err_ridge <- apply(pred_ridge, 2, function(col) mse(test_age, col)) 
 min_err_ridge <- min(err_ridge)
+
+##Save some results for later analysis 
+which_min_err <- which.min(err_ridge)
 pred_min_ridge <- pred_ridge[, which_min_err]
+
+##Return errors
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:575]:
```suggestion
## Return errors
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 14]
```diff
@@ -552,19 +555,34 @@ or correlated predictors. As we reduce the importance of one feature, we can
...
+This is much lower than the test error for the model without regularisation:
+
+```{r pred-ridge-lm-noreg}
+err_lm  
+```
+
+
 We can see where on the continuum of lambdas we've picked a model by plotting
 the coefficient paths again. In this case, we've picked a model with fairly
 modest shrinkage.
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:588]: Have we defined shrinkage? We did previously but I think that this might have been lost?

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 8 above)
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:783]: Alternate suggestion to Alan's above...

```suggestion
# Extract the coefficients from the model with the lowest mean squared error from cross-validation
```

File: _episodes_rmd/03-regression-regularisation.Rmd [hunk 15]
```diff
@@ -736,35 +755,45 @@ drawplot()
...
 lasso <- cv.glmnet(methyl_mat[, -1], age, alpha = 1)
 plot(lasso)
+
+##Extract the coefficients and return those that are non-zero
 coefl <- coef(lasso, lasso$lambda.min)
-selected_coefs <- as.matrix(coefl)[which(coefl != 0), 1]
+selected_coefs <- as.matrix(coefl)[which(coefl != 0), 1] ; selected_coefs
+```
+
+We can see that cross-validation has selected a value of $\lambda$ resulting in `r length(selected_coefs)-1` features and the intercept. We can cross-reference these features with the Hovarth data to compare the features to the Hovarth signature:
```
Comment (ailithewing) [inline _episodes_rmd/03-regression-regularisation.Rmd:788]:
```suggestion
We can see that cross-validation has selected a value of $\lambda$ resulting in `r length(selected_coefs)-1` features and the intercept. We can compare these features with the features included in the Horvath signature using the Hovarth data:
```

**Review by ailithewing — COMMENTED** [review]
I've made some minor suggestions on top of Alan's. Can merge after they're incorporated. Amazing work...such an involved episode!

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 15 above)
Comment (alanocallaghan) [reply _episodes_rmd/03-regression-regularisation.Rmd:788]: Is "by using the Hovarth data" adding anything here?

Comment (alanocallaghan) [reply _episodes_rmd/03-regression-regularisation.Rmd:788]: Also if referencing it, I'd suggest clarifying we're not really looking at the data, it's just coefficients from a previous analysis using roughly similar methods

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 6 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd:778]:
```suggestion
```{r lasso-cv, fig.cap="Cross-validated mean squared error for different values of lambda under a LASSO penalty.", fig.alt="Mean squared error (with confidence intervals) for a range of different lambda values. Mean squared error values are shown as red dots, with black error bars on the y-axis showing the corresponding confidence intervals."}
```

Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd:778]: added plural 'confidence intervals' for first mention and corresponding-> showing in the second sentence

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 9 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd:785]: This is so much nicer!

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 11 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd:558]: Agreed - this is so much clearer

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 14 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd]: "Shrinkage" isn't used but the episode refers to "shrinking of the coefficients" further up. Have changed "modest shrinkage" to "modest coefficient shrinking" in 96e6721

Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd]: Would you prefer shrinkage to be defined somewhere?

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 8 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd]: Changed in [f20b2ec](https://github.com/mallewellyn/high-dimensional-stats-r/commit/f20b2eca66b2d20d4e9d2438bdba61f5c9f506b3)

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 15 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd:792]: Could remove the Hovarth part if it's not really necessary. Would reduce the length of the lesson slightly also!

**Comment by mallewellyn** [comment]
Thanks Ailith! I've made most of the changes but have left a few comments above where unsure

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 14 above)
Comment (ailithewing) [reply _episodes_rmd/03-regression-regularisation.Rmd]: No I think this is fine.

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 15 above)
Comment (ailithewing) [reply _episodes_rmd/03-regression-regularisation.Rmd:792]: I think we could take it out.

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 6 above)
Comment (ailithewing) [reply _episodes_rmd/03-regression-regularisation.Rmd:778]: Looks good!

File: _episodes_rmd/03-regression-regularisation.Rmd (hunk 15 above)
Comment (mallewellyn) [reply _episodes_rmd/03-regression-regularisation.Rmd]: Removed in [33d7163](https://github.com/mallewellyn/high-dimensional-stats-r/commit/33d71635d668fbee0a84ed540c69486b90836b94)

Now I'm thinking that we could also remove the heatmap as I'm not sure what this adds/no text following:

https://github.com/mallewellyn/high-dimensional-stats-r/blob/33d71635d668fbee0a84ed540c69486b90836b94/_episodes_rmd/03-regression-regularisation.Rmd#L795-L817

**Review by ailithewing — APPROVED** [review]
